### PR TITLE
103778 - CMOS App expand disk

### PR DESCRIPTION
**Pull request by ritanorinho**
**Changes**:
- **Create Volume to Use Pipe in VM**: Added functionality to create a volume for using the pipe created in the virtual machine.
- **Resize Largest Physical Volume in VM**: Implemented logic to resize the largest physical volume in the virtual machine
- **Execute Button**: Added a button to execute the above actions.

### CODE REVIEW

File: mesinabox/kioskapp/views/index.html [hunk 1]
```diff
@@ -19,6 +19,7 @@ <h1 class="title-center">CMOS</h1>
         <div id="content"></div>
         <div id="enrollButtonDiv"></div>
         <button class="cmf-btn" id="ManageSslCertBtn">Manage SSL Certificate</button>
+        <button class="cmf-btn" id="expandDiskButton">Expand disk</button>
```
Comment (vitorpmoreira) [inline mesinabox/kioskapp/views/index.html:22]: Disk

File: mesinabox/kioskapp/server.js [hunk 2]
```diff
@@ -312,6 +312,18 @@ app.get('/enrollstart', (req, res) => {
   });
 });
 
+
+
+// API endpoint to execute script to expand disk space
+app.post('/expandDisk', (req, res) => {
+
+  const pipePath = "pipe";
+  const commandToRun = 'pvresize "$(pvdisplay --units b -c | awk -F: \'{print $1,$7}\' | sort -k2 -nr | head -n1 | awk \'{print $1}\')"';
```
Comment (vitorpmoreira) [inline mesinabox/kioskapp/server.js:321]: This is missing the commands to setup the disk as well. This feature must setup the disk, if it hasn't been setup yet, OR expand the disk if it was already setup.

File: mesinabox/kioskapp/server.js [hunk 3]
```diff
@@ -1,9 +1,9 @@
 const express = require('express');
+const { spawn, exec } = require('child_process');
```
Comment (vitorpmoreira) [inline mesinabox/kioskapp/server.js:2]: exec is not in use

File: mesinabox/kioskapp/server.js [hunk 4]
```diff
@@ -312,6 +312,18 @@ app.get('/enrollstart', (req, res) => {
...
+
+
+// API endpoint to execute script to expand disk space
+app.post('/expandDisk', (req, res) => {
+
+  const pipePath = "pipe";
+  const commandToRun = 'pvresize "$(pvdisplay --units b -c | awk -F: \'{print $1,$7}\' | sort -k2 -nr | head -n1 | awk \'{print $1}\')"';
+  const wstream = fs.createWriteStream(pipePath)
+  wstream.write(commandToRun)
+  wstream.close()
```
Comment (vitorpmoreira) [inline mesinabox/kioskapp/server.js:324]: we need to collect the exit code and show success or failure to the GUI

File: mesinabox/stack/nodetest-deployment.yaml [hunk 5]
```diff
@@ -32,9 +32,14 @@ spec:
...
           name: nodetest-volume
+        - name: pipe-volume
+          mountPath: /opt/app-root/src/pipe
       volumes:
         - name: nodetest-volume
           persistentVolumeClaim:
            claimName: nodetest-pvc
+        - name: pipe-volume
+          hostPath:
+            path: /etc/resizeDiskPipe            # Path on the host node containing the named pipe
```
Comment (vitorpmoreira) [inline mesinabox/stack/nodetest-deployment.yaml:43]: From the linux filesystem, it appears that the named pipe is better placed at the /var/run dir. https://stackoverflow.com/questions/15794696/what-is-the-proper-place-to-put-named-pipes-on-linux

**Review by vitorpmoreira — CHANGES_REQUESTED** [review]

File: mesinabox/kioskapp/views/index.html (hunk 1 above)
Comment (ritanorinho) [reply mesinabox/kioskapp/views/index.html:22]: done

File: mesinabox/kioskapp/server.js (hunk 3 above)
Comment (ritanorinho) [reply mesinabox/kioskapp/server.js:2]: removed

File: mesinabox/kioskapp/server.js (hunk 2 above)
Comment (ritanorinho) [reply mesinabox/kioskapp/server.js]: done

File: mesinabox/kioskapp/public/client.js [hunk 6]
```diff
@@ -1,4 +1,30 @@
 const clusterAddress = window.location.origin;
+const expandDiskButton = document.getElementById("ExpandDiskBtn");
+expandDiskButton.addEventListener("click", function() {
+    fetch('/expandDisk', {
+        method: 'POST',
+        headers: {
+            'Content-Type': 'application/json'
+        },
+        body: JSON.stringify({
```
Comment (vitorpmoreira) [inline mesinabox/kioskapp/public/client.js:9]: body is not required

File: mesinabox/kioskapp/public/client.js [hunk 7]
```diff
@@ -1,4 +1,30 @@
...
+            size: 5
+        })
+    })
+    .then(response => {
+        if (!response.ok) {
+            throw new Error('Network response was not ok');
```
Comment (vitorpmoreira) [inline mesinabox/kioskapp/public/client.js:15]: We need a better message here. Something like "Failed to expand the disk. Details: {error details here}"

File: mesinabox/kioskapp/server.js [hunk 8]
```diff
@@ -312,6 +312,62 @@ app.get('/enrollstart', (req, res) => {
...
+  fifoStream.close();
+
+    // Start listening for exit code
+
+    let timeout = 10000 //stop waiting after 10 seconds (something might be wrong)
+    const timeoutStart = Date.now()
+    const myLoop = setInterval(function () {
+        if (Date.now() - timeoutStart > timeout) {
+            clearInterval(myLoop);
+            console.log("timed out")
```
Comment (vitorpmoreira) [inline mesinabox/kioskapp/server.js:339]: we should send a timeout error to the gui to be presented

File: mesinabox/kioskapp/server.js [hunk 9]
```diff
@@ -312,6 +312,62 @@ app.get('/enrollstart', (req, res) => {
...
+        } else {
+            //if output.txt exists, read it
+            if (fs.existsSync(outputPath)) {
+                clearInterval(myLoop);
+                const data = fs.readFileSync(outputPath).toString()
+                if (fs.existsSync(outputPath)) fs.unlinkSync(outputPath) //delete the output file
+                res.send(data);
```
Comment (vitorpmoreira) [inline mesinabox/kioskapp/server.js:346]: if the output is an error, the res should send the data back with an error, probably 500 and the details of that to be presented in the gui

File: mesinabox/stack/nodetest-sc.yaml [hunk 10]
```diff
@@ -0,0 +1,11 @@
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+  name: nodetest
+parameters:
+  csi.storage.k8s.io/fstype: xfs
+  topolvm.io/device-class: externalDiskVG
+provisioner: topolvm.io
+reclaimPolicy: Delete
```
Comment (vitorpmoreira) [inline mesinabox/stack/nodetest-sc.yaml:9]: we should set as Retain, so that data from old volumes is not lost

**Review by vitorpmoreira — CHANGES_REQUESTED** [review]

File: mesinabox/kioskapp/public/client.js (hunk 6 above)
Comment (ritanorinho) [reply mesinabox/kioskapp/public/client.js:9]: done

File: mesinabox/kioskapp/public/client.js (hunk 7 above)
Comment (ritanorinho) [reply mesinabox/kioskapp/public/client.js:15]: done

File: mesinabox/kioskapp/public/client.js (hunk 6 above)
Comment (ritanorinho) [reply mesinabox/kioskapp/public/client.js:9]: done

File: mesinabox/kioskapp/public/client.js (hunk 7 above)
Comment (ritanorinho) [reply mesinabox/kioskapp/public/client.js:15]: done

**Review by vitorpmoreira — APPROVED** [review]

**Pull request merged**